### PR TITLE
Lots of Recipes

### PR DESCRIPTION
**Pull request by TwistedGate**
Contains recipes for:
Voltmeter, capacitor-backpack, toolbox, Engineer's Workbench, treated scaffold, metal scaffolds (and their variations), both reinforced and normal crate, all fences, engineering blocks (redstone, light, heavy), all coil blocks, generator and finaly radiator.

Edit: And also slight formatting for the already existing recipes aswell.

### CODE REVIEW

File: src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java [hunk 1]
```diff
@@ -107,77 +112,556 @@ protected void registerRecipes(@Nonnull Consumer<IFinishedRecipe> out)
...
+			.key('s', IETags.treatedStick)
+			.addCriterion("has_treated_planks", hasItem(IETags.getItemTag(IETags.treatedWood)))
+			.build(out);
+		ShapedRecipeBuilder.shapedRecipe(MetalDecoration.aluScaffolding.get(MetalScaffoldingType.STANDARD), 6)
+			.patternLine("iii")
+			.patternLine(" s ")
+			.patternLine("s s")
+			.key('i', IETags.getTagsFor(EnumMetals.ALUMINUM).ingot)
+			.key('s', IETags.aluminumRod)
+			.addCriterion("has_"+toPath(MetalDecoration.aluScaffolding.get(MetalScaffoldingType.STANDARD)), hasItem(MetalDecoration.aluScaffolding.get(MetalScaffoldingType.STANDARD)))
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java]: Shouldn't this be 2 calls to `hasCriterion` for alu ingots and rods? Unlocking recipes with the output item seems strange.

File: src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java [hunk 2]
```diff
@@ -107,77 +112,556 @@ protected void registerRecipes(@Nonnull Consumer<IFinishedRecipe> out)
...
+			.addCriterion("has_"+toPath(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD)), hasItem(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD)))
+			.build(out, toRL("steel_scaffolding_grate_top_from_standard"));
+		ShapelessRecipeBuilder.shapelessRecipe(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.WOODEN_TOP))
+			.addIngredient(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.GRATE_TOP))
+			.addCriterion("has_"+toPath(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.GRATE_TOP)), hasItem(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.GRATE_TOP)))
+			.build(out, toRL("steel_scaffolding_wooden_top_from_grate_top"));
+		ShapelessRecipeBuilder.shapelessRecipe(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD))
+			.addIngredient(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.WOODEN_TOP))
+			.addCriterion("has_"+toPath(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD)), hasItem(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD)))
+			.build(out, toRL("steel_scaffolding_standard_from_wooden_top"));
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java]: This block should probably be turned into a loop along the lines of
```java
int numScaffoldingTypes = MetalScaffoldingTypes.values().length;
for (MetalScaffoldingType from : MetalScaffoldingType.values()) {
    MetalScaffoldingType to = MetalScaffoldingType.values()[(from.ordinal()+1)%numScaffoldingTypes);
   //Add conversion recipes from `from` to `to`
}
```

File: src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java [hunk 3]
```diff
@@ -107,77 +112,556 @@ protected void registerRecipes(@Nonnull Consumer<IFinishedRecipe> out)
...
 
-		ShapedRecipeBuilder.shapedRecipe(StoneDecoration.alloybrick, 2).patternLine("sb").patternLine("bs").key('s', Tags.Items.SANDSTONE).key('b', Tags.Items.INGOTS_BRICK).addCriterion("has_brick", hasItem(Tags.Items.INGOTS_BRICK)).build(out);
+		ShapedRecipeBuilder.shapedRecipe(WoodenDecoration.treatedFence, 3)
+			.patternLine("isi")
+			.patternLine("isi")
+			.key('i', IETags.getItemTag(IETags.treatedWood))
+			.key('s', IETags.treatedStick)
+			.addCriterion("has_treated_planks", hasItem(IETags.getItemTag(IETags.treatedWood)))
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java]: Maybe add a second criterion on `has_treated_sticks`? (same for other recipes)

File: src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java (hunk 1 above)
Comment (TwistedGate) [reply src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java]: Oh whops, meant to change it befor commiting but forgot. Gonna work on that now.

**Comment by TwistedGate** [comment]
Anything else that would need to be done?

File: src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java [hunk 4]
```diff
@@ -107,77 +112,581 @@ protected void registerRecipes(@Nonnull Consumer<IFinishedRecipe> out)
...
+			MetalScaffoldingType from = MetalScaffoldingType.values()[i];
+			MetalScaffoldingType to = null;
+			if((i+1)>=MetalScaffoldingType.values().length) // Meant to prevent this whole thing to throw IndexOutOfBounds.
+			{
+				to = MetalScaffoldingType.STANDARD; // And to have wood_top -> standard conversion
+			}
+			else
+			{
+				to = MetalScaffoldingType.values()[i+1];
+			}
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java]: What I suggested in my comment shouldn't throw AIOOBEs, that's what the `%numScaffoldingTypes` is for. If you want to keep things this way you should remove the `{}`s for the if, in IE we don't use them for statements with a single-line body (I don't really agree with that style in general, but this feels like far too "big" vertically).

**Pull request merged**